engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/unboxing/Layout.java

```diff
@@ -0,0 +1,280 @@
+  public UnboxingAtom.FieldSetterNode getUncachedFieldSetter(int index) {
+    return uncachedFieldSetters[index];
+  }
+
+  public UnboxingAtom.FieldSetterNode buildSetter(int index) {
+    return fieldSetterFactories[index].createNode();
+  }
+
+  public boolean isDoubleAt(int fieldIndex) {
+    return Flags.isDoubleAt(inputFlags, fieldIndex);
+  }
+
+  public int arity() {
+    return fieldToStorage.length;
+  }
+
+  public int[] getFieldToStorage() {
+    return fieldToStorage;
+  }
+
+  public NodeFactory<? extends UnboxingAtom.InstantiatorNode> getInstantiatorFactory() {
+    return instantiatorFactory;
+  }
+
+  /**
+   * This acts as the main method of instantiating unboxing atoms throughout the system. It will try
+   * to reuse a known layout if it fits the data, or allocate a new one. The list of allocated
+   * layouts is stored in the constructor. This way we can better reuse them, allowing for less
+   * splitting throughout the system. This node will allocate these layouts, until a limit is
+   * reached, at which point it will fall back onto {@link AtomConstructor#getBoxedLayout()} for
+   * data that does not fit the known layouts.
+   */
+  public static class CreateUnboxedInstanceNode extends InstantiateNode.CreateInstanceNode {
+    @Child UnboxingAtom.DirectCreateLayoutInstanceNode boxedLayout;
+    @Children UnboxingAtom.DirectCreateLayoutInstanceNode[] unboxedLayouts;
+    private final int arity;
+    private @CompilerDirectives.CompilationFinal boolean constructorAtCapacity;
+
+    private final AtomConstructor constructor;
+
+    CreateUnboxedInstanceNode(AtomConstructor constructor) {
+      this.constructor = constructor;
+      this.arity = constructor.getArity();
+      this.boxedLayout =
+          new UnboxingAtom.DirectCreateLayoutInstanceNode(
+              constructor, constructor.getBoxedLayout());
+      unboxedLayouts = new UnboxingAtom.DirectCreateLayoutInstanceNode[0];
+      updateFromConstructor();
+    }
+
+    public static CreateUnboxedInstanceNode create(AtomConstructor constructor) {
+      return new CreateUnboxedInstanceNode(constructor);
+    }
+
+    @Override
+    @ExplodeLoop
+    public Object execute(Object[] arguments) {
+      var flags = computeFlags(arguments);
+      if (flags == 0) {
+        return boxedLayout.execute(arguments);
+      }
+      for (int i = 0; i < unboxedLayouts.length; i++) {
+        if (unboxedLayouts[i].layout.inputFlags == flags) {
+          return unboxedLayouts[i].execute(arguments);
+        }
+      }
+
+      if (!constructorAtCapacity) {
+        CompilerDirectives.transferToInterpreterAndInvalidate();
+        var lock = constructor.getLayoutsLock();
+        lock.lock();
+        try {
+          var layouts = constructor.getUnboxingLayouts();
+          if (layouts.length != this.unboxedLayouts.length) {
+            // Layouts changed since we last tried; Update & try again
+            updateFromConstructor();
+            return execute(arguments);
```

Comment:
`execute` shall probably _not_ be executed while holding the `lock`. Release the lock first, then call `execute`.